iconservice/fee/fee_engine.py

```diff
@@ -0,0 +1,844 @@
+                 icx_storage: 'IcxStorage',
+                 icx_engine: 'IcxEngine'):
+
+        self._deploy_storage = deploy_storage
+        self._fee_storage = fee_storage
+        self._icx_storage = icx_storage
+        self._icx_engine = icx_engine
+
+    def get_deposit_info(self,
+                         context: 'IconScoreContext',
+                         score_address: 'Address',
+                         block_height: int) -> DepositInfo:
+        """
+        Gets the SCORE deposit information
+
+        :param context: IconScoreContext
+        :param score_address: SCORE address
+        :param block_height: current block height
+        :return: score deposit information in dict
+                - SCORE Address
+                - Amount of issued total virtual step
+                - Amount of Used total virtual step
+                - deposits in list
+        """
+
+        self._check_score_valid(context, score_address)
+
+        deposit_meta = self._get_or_create_deposit_meta(context, score_address)
+
+        deposit_info = DepositInfo(score_address)
+
+        # Appends all deposits
+        for deposit in self._deposit_generator(context, deposit_meta.head_id):
+            deposit_info.deposits.append(deposit)
+
+            # Retrieves available virtual STEPs and deposits
+            if block_height < deposit.expires:
+                deposit_info.available_virtual_step += deposit.remaining_virtual_step
+                deposit_info.available_deposit += \
+                    max(deposit.remaining_deposit - deposit.min_remaining_deposit, 0)
+
+        return deposit_info
+
+    def add_deposit(self,
+                    context: 'IconScoreContext',
+                    tx_hash: bytes,
+                    sender: 'Address',
+                    score_address: 'Address',
+                    amount: int,
+                    block_height: int,
+                    term: int) -> None:
+        """
+        Deposits ICXs for the SCORE.
+        It may be issued the virtual STEPs for the SCORE to be able to pay share fees.
+
+        :param context: IconScoreContext
+        :param tx_hash: tx hash of the deposit transaction
+        :param sender: ICX sender
+        :param score_address: SCORE
+        :param amount: amount of ICXs in loop
+        :param block_height: current block height
+        :param term: deposit term in blocks
+        """
+        # [Sub Task]
+        # - Deposits ICX
+        # - Calculates Virtual Step
+        # - Updates Deposit Data
+
+        if not (self._MIN_DEPOSIT_AMOUNT <= amount <= self._MAX_DEPOSIT_AMOUNT):
+            raise InvalidRequestException('Invalid deposit amount')
+
+        if not (self._MIN_DEPOSIT_TERM <= term <= self._MAX_DEPOSIT_TERM):
+            raise InvalidRequestException('Invalid deposit term')
+
+        self._check_score_ownership(context, sender, score_address)
+
+        deposit_meta = self._get_or_create_deposit_meta(context, score_address)
```

Comment:
This line is redundant and can be removed. Newly created Deposit object will be inserted in the list in `_insert_deposit` later.